File: backend/src/main/java/com/tododuk/domain/label/service/LabelService.java
package com.tododuk.domain.label.service;

import com.tododuk.domain.label.entity.Label;
import com.tododuk.domain.label.repository.LabelRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@RequiredArgsConstructor
@Service
public class LabelService {

    private final LabelRepository labelRepository;

    @Transactional(readOnly = true)
    public long countLabels(){
        return labelRepository.count();
    }

    @Transactional(readOnly = true)
    public List<Label> getLabels(){
        return labelRepository.findAll();
    }

    @Transactional
    public Label createLabel(Label label) {
        labelRepository.save(label);
        return label;
    }

    @Transactional
    public Label createLabelIfNotExists(String name, String color) {
        Label label;
        if (!labelRepository.existsByName(name)) {
            label = createLabel(Label.builder()
                    .name(name)
                    .color(color)
                    .build());
        } else {
            label = labelRepository.findByName(name);
        }
        return label;
    }
}

File: backend/src/main/java/com/tododuk/domain/label/service/LabelService.kt
package com.tododuk.domain.label.service

import com.tododuk.domain.label.entity.Label
import com.tododuk.domain.label.repository.LabelRepository
import lombok.RequiredArgsConstructor
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@RequiredArgsConstructor
@Service
class LabelService(private val labelRepository: LabelRepository) {

    @Transactional(readOnly = true)
    fun countLabels(): Long {
        return labelRepository.count()
    }

    @get:Transactional(readOnly = true)
    val labels: MutableList<Label?>
        get() = labelRepository.findAll()

    @Transactional
    fun createLabel(label: Label): Label {
        labelRepository.save<Label?>(label)
        return label
    }

    @Transactional
    fun createLabelIfNotExists(name: String, color: String): Label {
        return labelRepository.findByName(name)
            ?: createLabel(Label(name=name, color=color))
    }
}

